Fix resource shared when duplicating an instanced scene

Rindbee opened the pull request:
Fix #45350, fix #47918.
Depends on #65011.

For resources with `resource_local_to_scene` enabled in the subscene, the resource is already set when the subscene is instantiated, so does not need to be set again.

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->

Rindbee committed:
Fix resource shared when duplicating an instanced scene

For resources with `resource_local_to_scene` enabled in the sub-scene,
the resource is already set when the sub-scene is instantiated, so does
not need to be set again. Just needs to update the property of the
resource according to the value in the main scene.

RandomShaper commented on scene/main/node.cpp:
```diff
@@ -2193,10 +2196,22 @@ Node *Node::_duplicate(int p_flags, HashMap<const Node *, Node *> *r_duplimap) c
+					HashMap<Ref<Resource>, Ref<Resource>> tmp;
+					current_res->configure_for_local_scene(node, tmp);
+					current_res->setup_local_to_scene();
+					continue;
+				}
+				if (E.usage & PROPERTY_USAGE_DO_NOT_SHARE_ON_DUPLICATE) {
 					current_node->set(name, res->duplicate());
+				} else {
+					current_node->set(name, value);
 				}
```
```suggestion
				} else {
					if (E.usage & PROPERTY_USAGE_DO_NOT_SHARE_ON_DUPLICATE) {
						current_node->set(name, res->duplicate());
					} else {
						current_node->set(name, value);
					}
				}
```

RandomShaper reviewed (changes requested):
Looks good. There's just the comment about structure to improve readability.

Rindbee commented:
There is also a similar situation here https://github.com/godotengine/godot/pull/64999#issuecomment-1229569123

Rindbee commented on scene/main/node.cpp:
```diff
@@ -2193,16 +2197,43 @@ Node *Node::_duplicate(int p_flags, HashMap<const Node *, Node *> *r_duplimap) c
+						if (current_res.is_valid() && current_res->is_local_to_scene()) {
+							// For local-to-scene resources in sub-scene, the old instance needs to be preserved to
+							// prevent sharing from being broken. Just update it with the data from the source resource.
+							current_res->copy_from(res, resources_local_to_sub_scene, true);
```
The modified `copy_from` in #65011 is used here

akien-mga commented:
The build seems to fail.

Rindbee commented:
Yes, here uses the `copy_from` after the modified in #65011.

RandomShaper commented:
@Rindbee, you may want to rebase this PR on top of the other one this depends on so CI build passes. When the other one is merged, this one will get automatically trimmed down to the only additional commit.

Rindbee commented:
I think there are fewer records about the correlation between resources between tscn.

All the related (merged and not yet merged) PRs are based on a judgment: Resources with local-to-scene enabled on the same property of the same object that recorded in the main scene, is  a copy of the source resource which recorded in the sub-scene. 

In most cases, this may be correct. But in fact, this situation cannot be ruled out: Really hope to use new resources to cover the resources on the scene.

AThousandShips commented on core/io/resource.cpp:
```diff
@@ -181,8 +181,39 @@ Error Resource::copy_from(const Ref<Resource> &p_resource) {
+						sub_res->copy_from(sub_res_of_from, p_remap_cache, true);
+						p_remap_cache[sub_res_of_from] = sub_res;
+						value = sub_res;
+					} else { // A copy of the source resource is required to overwrite the previous one
```
```suggestion
					} else { // A copy of the source resource is required to overwrite the previous one.
```
Code style nitpick, same elsewhere

KoBeWi commented on scene/main/node.cpp:
```diff
@@ -2513,14 +2517,36 @@ Node *Node::_duplicate(int p_flags, HashMap<const Node *, Node *> *r_duplimap) c
-				Resource *res = Object::cast_to<Resource>(value);
-				if (res) { // Duplicate only if it's a resource
-					current_node->set(name, res->duplicate());
+			Ref<Resource> res = value;
+			if (res.is_valid()) { // Duplicate only if it's a resource.
+				if (instantiated && res->is_local_to_scene()) {
+					// Find the shared copy of the source resource.
+					HashMap<Ref<Resource>, Ref<Resource>>::Iterator R = resources_local_to_sub_scene.find(res);
+					if (R) {
+						value = R->value;
+					} else {
+						Ref<Resource> current_res = current_node->get(name);
+						if (current_res.is_valid() && current_res->is_local_to_scene()) {
+							// Simply copy the data from the source resource to update the resource that was previously set.
+							current_res->copy_from(res, resources_local_to_sub_scene, true);
+							resources_local_to_sub_scene[res] = current_res;
+							value = current_res;
+						} else { // A copy of the source resource is required to overwrite the previous one.
+							Ref<Resource> dupe = res->duplicate_for_local_scene(node, resources_local_to_sub_scene);
+							resources_local_to_sub_scene[res] = dupe;
+							value = dupe;
+						}
+					}
+				} else if (E.usage & PROPERTY_USAGE_ALWAYS_DUPLICATE) {
+					value = res->duplicate();
 				}
+			}
```
It's the third time you are writing this code (first two were in #64487). Can it be moved to some method?

KoBeWi commented on core/io/resource.h:
```diff
@@ -81,13 +81,14 @@ class Resource : public RefCounted {
-	virtual void reset_state(); //for resources that use variable amount of properties, either via _validate_property or _get_property_list, this function needs to be implemented to correctly clear state
-	virtual Error copy_from(const Ref<Resource> &p_resource);
+	virtual void reset_state(); //For resources that use variable amount of properties, either via _validate_property or _get_property_list, this function needs to be implemented to correctly clear state.
+	static Ref<Resource> get_remap_resource(const Ref<Resource> &p_resource, HashMap<Ref<Resource>, Ref<Resource>> p_remap_cache, const Ref<Resource> p_fallback, Node *p_for_scene);
```
This method shouldn't be mixed in here. You could put it under the static methods above.

YuriSizov commented:
Your PR currently contains 2 commits with the same description.

KoBeWi commented:
#65011 should be merged first as this PR builds on top.

YuriSizov commented:
Yes, but the second commit here is still wrong 🙃

reduz commented on core/io/resource.h:
```diff
@@ -81,13 +81,14 @@ class Resource : public RefCounted {
-	virtual void reset_state(); //for resources that use variable amount of properties, either via _validate_property or _get_property_list, this function needs to be implemented to correctly clear state
-	virtual Error copy_from(const Ref<Resource> &p_resource);
+	virtual void reset_state(); //For resources that use variable amount of properties, either via _validate_property or _get_property_list, this function needs to be implemented to correctly clear state.
+	virtual Error copy_from(const Ref<Resource> &p_resource, HashMap<Ref<Resource>, Ref<Resource>> &remap_cache, bool p_keep_local_to_scene_resource_instance = false);
```
I would honestly make a different function like `copy_from_with_remap` or something like this, so other places that don't need to use this logic don't  get the code polluted unnecesarily. In fact, if this is only used in packed_scene, I would just make this a local function of packed_scene.

AThousandShips commented on core/io/resource.h:
```diff
@@ -81,13 +81,14 @@ class Resource : public RefCounted {
 	virtual bool editor_can_reload_from_file();
-	virtual void reset_state(); //for resources that use variable amount of properties, either via _validate_property or _get_property_list, this function needs to be implemented to correctly clear state
-	virtual Error copy_from(const Ref<Resource> &p_resource);
+	virtual void reset_state(); //For resources that use variable amount of properties, either via _validate_property or _get_property_list, this function needs to be implemented to correctly clear state.
```
```suggestion
	virtual void reset_state(); // For resources that use variable amount of properties, either via _validate_property or _get_property_list, this function needs to be implemented to correctly clear state.
```

AThousandShips commented on core/io/resource.h:
```diff
@@ -81,13 +81,14 @@ class Resource : public RefCounted {
-	static Node *(*_get_local_scene_func)(); //used by editor
-	static void (*_update_configuration_warning)(); //used by editor
+	static Node *(*_get_local_scene_func)(); //Used by editor.
+	static void (*_update_configuration_warning)(); //Used by editor.
```
```suggestion
	static void (*_update_configuration_warning)(); // Used by editor.
```

AThousandShips commented on core/io/resource.h:
```diff
@@ -81,13 +81,14 @@ class Resource : public RefCounted {
 public:
-	static Node *(*_get_local_scene_func)(); //used by editor
-	static void (*_update_configuration_warning)(); //used by editor
+	static Node *(*_get_local_scene_func)(); //Used by editor.
```
```suggestion
	static Node *(*_get_local_scene_func)(); // Used by editor.
```

reduz commented on scene/main/node.cpp:
```diff
@@ -2498,6 +2498,9 @@ Node *Node::_duplicate(int p_flags, HashMap<const Node *, Node *> *r_duplimap) c
 
 	for (List<const Node *>::Element *N = node_tree.front(); N; N = N->next()) {
 		Node *current_node = node->get_node(get_path_to(N->get()));
+		if (instantiated && node != current_node) {
```
This should only be the case if the scene has editable children disabled. If enabled, you need to do property duplication in the children.

Rindbee replied:
This case was indeed missed before.

reduz reviewed (commented):
Looks OK overall, but make sure to check the comment I left.

reduz commented:
Sorry, I misunderstood the PR, as this is only the change in node. In this case, Its as I mentioned, you have to duplicate the children nodes manually if they have editable children flag set.

Rindbee commented:
Okay, the cases of enabling/disabling **Editable Children** has been completed for scene initialization and node duplication. I think it's ready to be reviewed again.

Rindbee committed:
Supplement the case of scene instantiation for "Editable Children"

This is a follow-up to #65011.

For scenes with **Editable Children** enabled, the main scene will record
more information and resource mapping will be valid for multiple nodes.

Rindbee commented:
`Node::_duplicate()` does not seem to be a suitable place to remap resources. It seems more appropriate to modify the mapping logic in `SceneTreeDock::paste_nodes()`.

I'll try to modify the mapping logic there later.

KoBeWi reviewed (approved):
Gave it some testing and it seems to work correctly.

Might need core re-review after latest changes.

Rindbee commented on editor/docks/scene_tree_dock.cpp:
```diff
@@ -956,6 +958,14 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
+						R.value->setup_local_to_scene();
+					}
+				}
+			}
```
This ensures that local-to-scene resources in newly copied child scene instances are effective when **Editable Children** is enabled.

The cache is now only cleared only when a new copy begins. As users edit the scene, it may lead to the use of released node instances as keys.

Repiteo commented:
Thanks!
